crates/pipeline-manager/src/compiler/rust_compiler.rs

```diff
@@ -585,92 +488,145 @@ fn main() {
+        "feldera_pipe_pipeline_{}_main",
+        pipeline_id.to_string().replace("-", "_")
+    );
+    let pipeline_globals_crate = format!(
+        "feldera_pipe_pipeline_{}_globals",
+        pipeline_id.to_string().replace("-", "_")
+    );
 
-    // Cargo.toml: Feldera crate locations
-    project_toml = project_toml
-        .replace(
-            "dbsp_adapters = { path = \"../../crates/adapters\", default-features = false }",
-            {
-                #[cfg(not(feature = "feldera-enterprise"))]
-                { r#"dbsp_adapters = { path = "../../crates/adapters" }"# }
-                #[cfg(feature = "feldera-enterprise")]
-                {
-                    r#"dbsp_adapters = { path = "../../crates/adapters", features = ["feldera-enterprise"] }
-dbsp-enterprise = { path = "../../crates/dbsp-enterprise" }"#
-                }
-            },
-        )
-        .replace(
-            "../../crates",
-            &format!("{}/crates", &config.dbsp_override_path),
-        )
-        .replace(
-            "../lib",
-            &format!(
-                "{}",
-                Path::new(&config.sql_compiler_home).join("lib").display()
-            ),
-        );
+    // Validate the crates
+    if !list_crate_names.contains(&pipeline_main_crate) {
+        return Err(RustCompilationError::SystemError(format!(
+            "Missing required main crate: '{pipeline_main_crate}'"
+        )));
+    }
+    if !list_crate_names.contains(&pipeline_globals_crate) {
+        return Err(RustCompilationError::SystemError(format!(
+            "Missing required globals crate: '{pipeline_globals_crate}'"
+        )));
+    }
+    for crate_name in &list_crate_names {
+        if !crate_name.starts_with("feldera_pipe_") {
+            return Err(RustCompilationError::SystemError(format!(
+                "Unknown crate '{}' not starting with 'feldera_pipe_'",
+                crate_name
+            )));
+        }
+    }
 
-    // Cargo.toml: add UDF dependencies
-    project_toml = project_toml.replace(
+    // Pipeline main crate: src/lib.rs -> src/main.rs
+    // ----------------------------------------------
+    // Converted into a Rust file which specifies a main instead, the file is renamed.
+    let lib_rs_path = extract_dir
+        .join(&pipeline_main_crate)
+        .join("src")
+        .join("lib.rs");
+    let main_rs_path = extract_dir
+        .join(&pipeline_main_crate)
+        .join("src")
+        .join("main.rs");
+    let mut main_rs_content = read_file_content(&lib_rs_path).await?;
+    main_rs_content.push_str(MAIN_FUNCTION);
+    #[cfg(feature = "feldera-enterprise")]
+    main_rs_content.push_str("\nextern crate dbsp_enterprise;");
+    create_new_file_with_content(&main_rs_path, &main_rs_content).await?;
+    fs::remove_file(&lib_rs_path)
+        .await
+        .map_err(|e| UtilError::IoError(format!("removing file '{}'", lib_rs_path.display()), e))?;
+
+    // Pipeline main crate: Cargo.toml
```

Comment:
we can have the compiler generate this from the start

Reply:
Sure, that would be handy for `src/lib.rs -> src/main.rs` and `Cargo.toml`.

For the former will still likely have to make a small edit for the enterprise external crate.

Reply:
(or that could also be a flag)

Reply:
sure